pkg/controller/license/trial/trial_controller.go

```diff
@@ -80,85 +78,95 @@ func (r *ReconcileTrials) Reconcile(request reconcile.Request) (reconcile.Result
 		return reconcile.Result{}, licensing.UpdateEnterpriseLicense(r, secret, license)
 	}
 
-	// 1. fetch trial status secret
-	var trialStatus corev1.Secret
-	err = r.Get(types.NamespacedName{Namespace: r.operatorNamespace, Name: licensing.TrialStatusSecretKey}, &trialStatus)
-	if errors.IsNotFound(err) {
-		// 2. if not present create one
-		err := r.initTrial(secret, license)
-		if err != nil {
-			return reconcile.Result{}, pkgerrors.Wrap(err, "failed to init trial")
-		}
-		return reconcile.Result{}, nil
+	// 1. reconcile trial status secret
+	if err := r.reconcileTrialStatus(request.NamespacedName); err != nil {
+		return reconcile.Result{}, err
 	}
-	if err != nil {
-		return reconcile.Result{}, pkgerrors.Wrap(err, "failed to retrieve trial status")
-	}
-	// 3. reconcile trial status
-	if err := r.reconcileTrialStatus(trialStatus); err != nil {
-		return reconcile.Result{}, pkgerrors.Wrap(err, "failed to reconcile trial status")
-	}
-	// 4. update trial secret if invalid to give user feedback
+
+	// 2. reconcile the trial license itself
 	trialSecretPopulated := license.IsMissingFields() == nil
-	if r.isTrialRunning() && trialSecretPopulated {
-		verifier := licensing.Verifier{
-			PublicKey: r.trialPubKey,
+	switch {
+	case r.trialState.IsTrialRunning() && !trialSecretPopulated:
+		// if the trial license fields are not populated at this point a user is trying to start a trial a second time
+		// with an empty trial secret, which is not a supported use case.
+		setValidationMsg(&secret, trialOnlyOnceMsg)
+	case !trialSecretPopulated && r.trialState.IsTrialActivationInProgress():
```

Comment:
Having a single `trialState.IsStarted()` instead of `trialState.IsTrialRunning()` and `trialState.IsTrialActivationInProgress()` would simplify the reading a little bit?

I think the switch would read better with more cases, consistent conditions ordering, and a return statement everywhere:

```go

switch {
    case !trialSecretPopulated && trialState.IsStarted():
        // user wants to start a trial a 2nd time
        return invalidOperation(secret, trialOnlyOnceMsg)
    case !trialSecretPopulated && !trialState.IsStarted():
        // user wants to init a trial for the first time
        return initTrial(secret)
     case trialSecretPopulated && !validLicense(license):
        // the existing license is invalid (expired or tampered)
        return invalidOperation(secret, invalidLicenseMsg(license))
    case trialSecretPopulated && validLicense(license) && !trialState.IsStarted():
        // valid license, let's consider the trial as started
        return trialState.SetStarted()
    case trialSecretPopulated && validLicense(license) && trialState.IsStarted():
        // all good, nothing to do
        return nil
}
```

Reply:
Much better!